src/hb-ot-layout-gsubgpos.hh

```diff
@@ -1774,8 +1874,46 @@ struct ContextFormat2
   bool subset (hb_subset_context_t *c) const
   {
     TRACE_SUBSET (this);
-    // TODO(subset)
-    return_trace (false);
+    auto *out = c->serializer->start_embed (*this);
+    if (unlikely (!c->serializer->extend_min (out))) return_trace (false);
+    out->format = format;
+    if (unlikely (!out->coverage.serialize_subset (c, coverage, this)))
+      return_trace (false);
+
+    hb_map_t klass_map;
+    out->classDef.serialize_subset (c, classDef, this, &klass_map);
+
+    hb_vector_t<unsigned> rulesets;
+    const hb_map_t *lookup_map = c->table_tag == HB_OT_TAG_GSUB ? c->plan->gsub_lookups : c->plan->gpos_lookups;
+    bool ret = true;
+    for (const hb_pair_t<unsigned, const OffsetTo<RuleSet>&>& _ : + hb_enumerate (ruleSet))
+    {
```

Comment:
Can you add a comment that explains why you need to retain rules that are otherwise subsetted out. From my understanding this is because we need an entry for each class in the subsetted class map.

Reply:
oh, my bad, we should only keep rulesets for class numbers that survived.